File: 01-beginner/03-services-and-actions/demo/python/add_two_ints_client.py
#!/usr/bin/env python3
"""Calls the add_two_ints service once, asynchronously, and prints the result."""

import sys

import rclpy
from rclpy.node import Node
from example_interfaces.srv import AddTwoInts


class AddTwoIntsClient(Node):
    def __init__(self):
        super().__init__("add_two_ints_client")
        self._client = self.create_client(AddTwoInts, "add_two_ints")

        # wait_for_service blocks (with a timeout, polled in a loop here)
        # until a server is actually advertising this service name. Without
        # this, calling the service before a server exists would just fail
        # — better to wait and tell the user what's happening.
        while not self._client.wait_for_service(timeout_sec=1.0):
            self.get_logger().info("Waiting for add_two_ints service...")

    def add(self, a: int, b: int) -> int:
        request = AddTwoInts.Request()
        request.a = a
        request.b = b

        # call_async returns a Future immediately instead of blocking this
        # function — we then hand control back to spin (via
        # spin_until_future_complete) so the executor can process the
        # response when it arrives. This is the pattern DEEP_DIVE.md
        # recommends over a blocking synchronous call, even in a simple
        # script like this one, so the habit carries over to contexts
        # where a synchronous call would deadlock.
        future = self._client.call_async(request)
        rclpy.spin_until_future_complete(self, future)
        return future.result().sum


def main():
    # Accept two integers from the command line, defaulting to 2 and 3 so
    # the demo works with no arguments too.
    a = int(sys.argv[1]) if len(sys.argv) > 1 else 2
    b = int(sys.argv[2]) if len(sys.argv) > 2 else 3

    rclpy.init()
    node = AddTwoIntsClient()
    result = node.add(a, b)
    node.get_logger().info(f"Result: {a} + {b} = {result}")
    node.destroy_node()
    rclpy.shutdown()


if __name__ == "__main__":
    main()

File: 01-beginner/03-services-and-actions/demo/cpp/src/add_two_ints_client.cpp
// Calls the add_two_ints service once, asynchronously, and prints the
// result. Takes two integers as command-line arguments (defaults 2, 3).
// C++ equivalent of demo/python/add_two_ints_client.py.

#include <chrono>
#include <cstdlib>
#include <memory>

#include "rclcpp/rclcpp.hpp"
#include "example_interfaces/srv/add_two_ints.hpp"

using AddTwoInts = example_interfaces::srv::AddTwoInts;
using namespace std::chrono_literals;

int main(int argc, char * argv[])
{
  rclcpp::init(argc, argv);

  auto node = rclcpp::Node::make_shared("add_two_ints_client");
  auto client = node->create_client<AddTwoInts>("add_two_ints");

  // Parse optional command-line args, same defaults as the Python client.
  long a = argc > 1 ? std::atol(argv[1]) : 2;
  long b = argc > 2 ? std::atol(argv[2]) : 3;

  // wait_for_service blocks (polled here in 1-second slices) until a
  // server advertises this service name, same reasoning as the Python
  // version: fail loudly and informatively rather than calling into a
  // service that doesn't exist yet.
  while (!client->wait_for_service(1s)) {
    if (!rclcpp::ok()) {
      RCLCPP_ERROR(node->get_logger(), "Interrupted while waiting for service");
      return 1;
    }
    RCLCPP_INFO(node->get_logger(), "Waiting for add_two_ints service...");
  }

  auto request = std::make_shared<AddTwoInts::Request>();
  request->a = a;
  request->b = b;

  // async_send_request returns a future; spin_until_future_complete blocks
  // this thread (safe here since main() isn't itself a callback) while
  // letting the executor process the response when it arrives.
  auto future = client->async_send_request(request);
  if (rclcpp::spin_until_future_complete(node, future) ==
    rclcpp::FutureReturnCode::SUCCESS)
  {
    RCLCPP_INFO(node->get_logger(), "Result: %ld + %ld = %ld",
      a, b, future.get()->sum);
  } else {
    RCLCPP_ERROR(node->get_logger(), "Service call failed");
  }

  rclcpp::shutdown();
  return 0;
}
